---
title: "LIHC data exploration"
output: 
    html_notebook: default
    html_document: default
---

# Set up

```{r}
library(tidyverse)
library(ggalluvial)
library(ComplexHeatmap)
```

```{r}
# Specify path to directory with downloaded data
data_dir_path = "./data/liver/"
stopifnot(file.exists(file.path(data_dir_path, 'exp')) & file.exists(file.path(data_dir_path, 'methy')) & 
          file.exists(file.path(data_dir_path, 'mirna')))
```

```{r}
# Inrease plot resolution in this notebook
options(repr.plot.res=200)
```

# Liver Hepatocellular Carcinoma (LIHC) Dataset

## (1) Closer look at the data 

```{r}
# load gene expression data as data.frame
data_exp <- read.table(file.path(data_dir_path, 'exp'))
dim(data_exp)
head(data_exp)
```

```{r}
# load methylation data as data.frame
data_met <- read.table(file.path(data_dir_path, 'methy'))
dim(data_met)
head(data_met)
```

```{r}
# load microRNA expression data as data.frame
data_mirna <- read.table(file.path(data_dir_path, 'mirna'))
dim(data_mirna)
head(data_mirna)
```

```{r}
tibble(sample = c(colnames(data_exp), colnames(data_met), colnames(data_mirna))) %>%
    distinct() %>%
    separate(sample, c('tcga', 'tss', 'participant', 'st'), sep="\\.") %>%
    group_by(st) %>%
    summarise(nsample=n())
```

So, we have cancer types 01 and 02, and some controls (11). **TODO: probably, we need to leave only 01 as only 2 samples for 02? Do we need to leave controls?**

## (2) TCGA Subtypes

How many unique samples are in our dataset? 

```{r}
tibble(sample = colnames(data_exp), layer="exp") %>%
    full_join(tibble(sample = colnames(data_met), layer="met"), by = c("sample", "layer")) %>%
    full_join(tibble(sample = colnames(data_mirna), layer="mirna"), by = c("sample", "layer")) %>%
    spread(layer,layer) %>%
    mutate(layers=paste(exp, met, mirna, sep=",")) %>%
    group_by(layers) %>%
    summarise(nsample=n())
```

Inspect the file with **known TCGA subtypes** (*LIHC_subtypes.tsv* in the *data/TCGA_subtypes* directory): 

```{r}
subtypes <- read_tsv("./data/TCGA_subtypes/LIHC_subtypes.tsv", show_col_types = FALSE)
dim(subtypes)
head(subtypes)
```

```{r}
subtypes %>%
    group_by(Subtype_miRNA, Subtype_mRNA, Subtype_DNAmeth, Subtype_iCluster, Subtype_Paradigm) %>%
    summarise(nsamples=n()) %>%
    ggplot(aes(y=nsamples, axis1=Subtype_miRNA, axis2=Subtype_mRNA, axis3=Subtype_DNAmeth, 
               axis4=Subtype_iCluster, axis5=Subtype_Paradigm)) +
      geom_alluvium(aes(fill=Subtype_iCluster), width = 0, knot.pos = 0, reverse = FALSE) +
      guides(fill = FALSE) +
      geom_stratum(width = 1/3, reverse = FALSE) +
      geom_text(stat = "stratum", aes(label = after_stat(stratum)),
                reverse = FALSE) +
      scale_x_continuous(breaks = 1:5, 
                         labels = c("Subtype miRNA", "Subtype mRNA", "Subtype DNAmeth",
                                    "Subtype iCluster", "Subtype Paradigm")) +
      labs(title="TCGA LIHC Subtypes", y="#samples")
```

- A lot of data with missing subtypes.
- Subtypes for mRNA, miRNA, DNAmeth and iCluster integrative classification are widely different. Paradigm and iCluster subtypes are somewhat similar, but not identical though.

## (3) Examine separate datatypes

### I. Gene expression

Log-transform the data and remove non-informative features. Draw transformed counts vs raw counts.

```{r}
preprocess.matrix <- function(data.matrix) {
  num = log2(data.matrix + 1)
  should.keep = (apply(num, 1, function(x) sd(x, na.rm=TRUE)) != 0) # we remove non-informative features 
  return(num[should.keep,])
}
dim(data_exp)                       
data_exp_tr <- preprocess.matrix(data_exp)
dim(data_exp_tr)
```

```{r}
tibble(feature_mean=c(rowMeans(data_exp, na.rm = TRUE), 
                      rowMeans(data_exp_tr, na.rm = TRUE)), 
       feature_sd=c(apply(data_exp, 1, function(x) sd(x, na.rm=TRUE)), 
                    apply(data_exp_tr, 1, function(x) sd(x, na.rm=TRUE))),
        data_matrix=c(rep("Raw read counts", dim(data_exp)[1]), 
                      rep("log(x+1) transformed counts", dim(data_exp_tr)[1]))) %>%
    mutate(data_matrix=as.factor(data_matrix)) %>%
    ggplot() + 
        geom_point(aes(x=feature_mean, y=feature_sd, size=1, alpha=0.1)) + 
        facet_wrap(data_matrix~., scales="free", ncol=1) + 
        guides(size=FALSE, alpha=FALSE) +
        ggtitle("Gene expression data")
```

```{r}
# Save transformed gene expression data (we are going to use it later)
write.table(data_exp_tr, file = file.path(data_dir_path, "exp_tr"), 
            row.names = TRUE, col.names = TRUE)
```

No heatmaps, as it takes ages to draw them.

### II. microRNA expression

As with gene expression, we apply the log(x+1) transformation to raw counts.

```{r}
dim(data_mirna)                       
data_mirna_tr <- preprocess.matrix(data_mirna)
dim(data_mirna_tr)
```

```{r}
tibble(feature_mean=c(rowMeans(data_mirna, na.rm = TRUE), 
                      rowMeans(data_mirna_tr, na.rm = TRUE)), 
       feature_sd=c(apply(data_mirna, 1, function(x) sd(x, na.rm=TRUE)),
                    apply(data_mirna_tr, 1, function(x) sd(x, na.rm=TRUE))),
        data_matrix=c(rep("Raw read counts", dim(data_mirna)[1]), 
                      rep("log(x+1) transformed counts", dim(data_mirna_tr)[1]))) %>%
    mutate(data_matrix=as.factor(data_matrix)) %>%
    ggplot() + 
        geom_point(aes(x=feature_mean, y=feature_sd, size=1, alpha=0.1)) + 
        facet_wrap(data_matrix~., scales="free", ncol=1) + 
        guides(size=FALSE, alpha=FALSE) +
        ggtitle("microRNA expression data")
```

```{r}
#Save transformed microRNA expression data (we are going to use it later)
write.table(data_mirna_tr, file = file.path(data_dir_path, "mirna_tr"), 
            row.names = TRUE, col.names = TRUE)
```

### III. DNA methylation

```{r}
hist(as.vector(as.matrix(data_met)), main= "DNA methylation distribution", xlab="beta values")
```




